ddl/reorg.go

```diff
@@ -174,15 +182,129 @@ type reorgInfo struct {
 	first  bool
 }
 
+func constructDescTableScanPB(tblInfo *model.TableInfo, pbColumnInfos []*tipb.ColumnInfo) *tipb.Executor {
+	tblScan := &tipb.TableScan{
+		TableId: tblInfo.ID,
+		Columns: pbColumnInfos,
+		Desc:    true,
+	}
+
+	return &tipb.Executor{Tp: tipb.ExecType_TypeTableScan, TblScan: tblScan}
+}
+
+func constructLimitPB(count uint64) *tipb.Executor {
+	limitExec := &tipb.Limit{
+		Limit: count,
+	}
+	return &tipb.Executor{Tp: tipb.ExecType_TypeLimit, Limit: limitExec}
+}
+
+func buildDescTableScanDAG(startTS uint64, tblInfo *model.TableInfo, columns []*model.ColumnInfo, limit uint64) (*tipb.DAGRequest, error) {
+	dagReq := &tipb.DAGRequest{}
+	dagReq.StartTs = startTS
+	_, timeZoneOffset := time.Now().In(time.UTC).Zone()
+	dagReq.TimeZoneOffset = int64(timeZoneOffset)
+	for i := range columns {
+		dagReq.OutputOffsets = append(dagReq.OutputOffsets, uint32(i))
+	}
+	dagReq.Flags |= model.FlagInSelectStmt
+
+	pbColumnInfos := model.ColumnsToProto(columns, tblInfo.PKIsHandle)
+	tblScanExec := constructDescTableScanPB(tblInfo, pbColumnInfos)
+	dagReq.Executors = append(dagReq.Executors, tblScanExec)
+	dagReq.Executors = append(dagReq.Executors, constructLimitPB(limit))
+	return dagReq, nil
+}
+
+func getColumnsTypes(columns []*model.ColumnInfo) []*types.FieldType {
+	colTypes := make([]*types.FieldType, 0, len(columns))
+	for _, col := range columns {
+		colTypes = append(colTypes, &col.FieldType)
+	}
+	return colTypes
+}
+
+// builds a desc table scan upon tblInfo.
+func (d *ddl) buildDescTableScan(ctx context.Context, startTS uint64, tblInfo *model.TableInfo, columns []*model.ColumnInfo, limit uint64) (distsql.SelectResult, error) {
+	dagPB, err := buildDescTableScanDAG(startTS, tblInfo, columns, limit)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	ranges := ranger.FullIntRange(false)
+	var builder distsql.RequestBuilder
+	builder.SetTableRanges(tblInfo.ID, ranges, nil).
+		SetDAGRequest(dagPB).
+		SetKeepOrder(true).
+		SetConcurrency(1)
+
+	builder.Request.NotFillCache = true
+	builder.Request.Priority = kv.PriorityLow
+	builder.Request.IsolationLevel = kv.SI
+
+	kvReq, err := builder.Build()
+	sctx := d.newContext()
+	result, err := distsql.Select(ctx, sctx, kvReq, getColumnsTypes(columns), statistics.NewQueryFeedback(0, nil, 0, false))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	result.Fetch(ctx)
+	return result, nil
+}
+
+// GetTableMaxRowID gets the last row id of the table.
+func (d *ddl) GetTableMaxRowID(startTS uint64, tblInfo *model.TableInfo) (maxRowID int64, emptyTable bool, err error) {
+	maxRowID = int64(math.MaxInt64)
+	var columns []*model.ColumnInfo
+	if tblInfo.PKIsHandle {
+		for _, col := range tblInfo.Columns {
+			if mysql.HasPriKeyFlag(col.Flag) {
+				columns = []*model.ColumnInfo{col}
+				break
+			}
+		}
+	} else {
+		columns = []*model.ColumnInfo{model.NewExtraHandleColInfo()}
+	}
+
+	if len(columns) == 0 {
+		return maxRowID, false, errors.Errorf("columns should not be nil")
```

Comment:
Update this error as a terror.

Reply:
And I think this case can't happen.

Reply:
So I think it is not necessary to be a terror.

Reply:
If you think so，why add this code？